### PR TITLE
Improve observability

### PR DESCRIPTION
**Pull request by kyscott18**

### CODE REVIEW

File: packages/core/src/common/metrics.ts [hunk 1]
```diff
@@ -20,24 +20,24 @@ export class MetricsService {
 
   ponder_indexing_total_seconds: prometheus.Gauge;
   ponder_indexing_completed_seconds: prometheus.Gauge;
-  ponder_indexing_completed_events: prometheus.Gauge<"network" | "event">;
+  ponder_indexing_completed_events: prometheus.Gauge<"event">;
```
Comment (0xOlias) [inline packages/core/src/common/metrics.ts:23]: We'll need a changeset for the metric label changes - could update the one we already have on `0.6`

File: packages/core/src/common/metrics.ts [hunk 2]
```diff
@@ -20,24 +20,24 @@ export class MetricsService {
...
   ponder_indexing_has_error: prometheus.Gauge;
 
-  ponder_indexing_function_duration: prometheus.Histogram<"network" | "event">;
-  ponder_indexing_function_error_total: prometheus.Counter<"network" | "event">;
-
+  ponder_indexing_function_duration: prometheus.Histogram<"event">;
   ponder_indexing_abi_decoding_duration: prometheus.Histogram;
 
+  ponder_sync_block: prometheus.Gauge<"network">;
+  // ponder_sync_events: prometheus.Gauge<"network">;
```
Comment (0xOlias) [inline packages/core/src/common/metrics.ts:32]: Remove?

File: packages/core/src/common/metrics.ts [hunk 3]
```diff
@@ -255,72 +260,87 @@ export async function getHistoricalSyncProgress(metrics: MetricsService) {
...
+  const extractMetric = (
+    metric: prometheus.MetricObjectWithValues<
+      prometheus.MetricValue<"network">
+    >,
+    network: string,
+  ) => {
+    return metric.values.find((m) => m.labels.network === network)?.value;
+  };
+
+  const totalBlocksMetric = await metrics.ponder_historical_total_blocks.get();
```
Comment (0xOlias) [inline packages/core/src/common/metrics.ts:272]: It's a pain that all of these have to be async when we know that there is no async `collect` function registered for any of our metrics. As a nerdsnipe, we could create a `SyncRegistry` subclass that enforces sync for everything.

File: packages/core/src/common/metrics.ts [hunk 4]
```diff
@@ -255,72 +260,87 @@ export async function getHistoricalSyncProgress(metrics: MetricsService) {
...
+      rps[networkName]!.shift();
+    }
+  }
 
-  return {
-    overall: progress,
-    networks,
-  };
+  return totalBlocksMetric.values.map(({ value, labels }) => {
+    const totalBlocks = value;
```
Comment (0xOlias) [inline packages/core/src/common/metrics.ts:305]: Could make this more compact by doing `const network = labels.network as string` here at the top

File: packages/core/src/ui/app.tsx [hunk 5]
```diff
@@ -31,13 +22,13 @@ export type UiState = {
...
     events: {
       eventName: string;
-      networkName: string;
       count: number;
       averageDuration: number;
-      errorCount: number;
     }[];
   };
 
+  app: Awaited<ReturnType<typeof getAppProgress>>;
```
Comment (0xOlias) [inline packages/core/src/ui/app.tsx:30]: Any reason not to do this trick for the `indexing` type right above?

File: packages/core/src/ui/app.tsx [hunk 6]
```diff
@@ -31,13 +22,13 @@ export type UiState = {
...
+
   realtimeSyncNetworks: {
```
Comment (0xOlias) [inline packages/core/src/ui/app.tsx:32]: Should be able to remove this from the type and initial object

**Review by 0xOlias — APPROVED** [review]
Looks great

File: packages/core/src/sync/index.ts [hunk 7]
```diff
@@ -455,6 +474,22 @@ export const createSync = async (args: CreateSyncParameters): Promise<Sync> => {
 
             yield { events, checkpoint: to };
             from = cursor;
+
+            getAppProgress(args.common.metrics).then(({ eta, progress }) => {
```
[the remark this answers is not in the document]
Comment (kyscott18) [reply packages/core/src/sync/index.ts:478]: I thought it was a bad idea to have metrics be a part of the hot loop

File: packages/core/src/sync/index.ts [hunk 8]
```diff
@@ -597,7 +638,16 @@ export const createSync = async (args: CreateSyncParameters): Promise<Sync> => {
             for (const network of args.networks) {
               updateRealtimeStatus({ checkpoint: cursor, network });
             }
-            args.onRealtimeEvent({ type: "block", checkpoint: to, events });
+            args
```
[the remark this answers is not in the document]
Comment (kyscott18) [reply packages/core/src/sync/index.ts:641]: This makes sure the log gets shown after the event is completed

File: packages/core/src/sync/index.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (0xOlias) [reply packages/core/src/sync/index.ts:478]: Here's the code that's actually running when we await the metrics: https://github.com/siimon/prom-client/blob/master/lib/histogram.js#L102-L125

In practice, we don't use the `collect` function pattern for any of these metrics (only use it for the postgres pool metrics which aren't used internally, and even those are actually synchronous).

**Pull request merged**